### PR TITLE
Make Allocation::bytes private

### LINKED ISSUE #62931: Allocation::bytes should be private
The `bytes` field of an [`Allocation`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/interpret/struct.Allocation.html) is a very subtle one -- the values in there often do not mean what one might intuitively think. First of all the doc comment should probably be improved, but secondly I think we need to make this field private. See https://github.com/rust-lang/rust/pull/62655#discussion_r306753992 for a very subtle bug that happened because it was not private.

Cc @oli-obk @eddyb

### PR DESCRIPTION
**Pull request by 197g**
Fixes #62931.

Direct immutable access to the bytes is still possible but redirected through the new method `raw_bytes_with_undef_and_ptr`, similar to `get_bytes_with_undef_and_ptr` but without requiring an interpretation context and not doing *any* relocation or bounds checks. The `size` of the allocation is stored separately which makes access as `Size` and `usize` more ergonomic.

cc: @RalfJung

### CODE REVIEW

File: src/librustc/mir/interpret/allocation.rs
Comment (Centril) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    crate bytes: Vec<u8>,
```

File: src/librustc/mir/interpret/allocation.rs
Comment (Centril) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    crate undef_mask: UndefMask,
```

File: src/librustc/mir/interpret/allocation.rs
Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Requires a customized implementation in `src/librustc/ich/impls_ty.rs` where destructuring assignment could no longer capture all fields. Would that be fine?

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: `crate` and `pub(crate)` are equivalent. Why would anything about that impl change?

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Oh. Right. My bad, I'm just not used to `crate field` syntax.

**Comment by HeroicKatora** [comment]
Changed `pub(crate)` to `crate` by amending the first commit.

**Comment by RalfJung** [comment]
I edited your OP to say "Fixes: #issue"; that will make GitHub auto-close it when the PR gets merged.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: Can't we restrict this further? Preferrably these would be private to this module only.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    /// Can be bigger than `bytes.len()` because we do not allocate all bytes immediately.
```

I don't understand what the second sentence here is telling me, is this a correct interpretation.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: Uh, given the PR title I had hoped this was just refactoring.  Please make sure your title reflects what you did!  Seems like beyond making `byte` private, this also changes the internal representaion?

This is two changes in one. Is there any chance you can split those into separate PRs? That would make them *a lot* easier to review.

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: This just moves the compression from `interpret/memory.rs` to `allocation`. It doesn't (at least shouldn't) change its implementation. It's hard to hide `undef_mask` without moving this since `memory` accesses it extensively. Well, there could be an intermediate PR with `fn undef_mask_mut(&mut self)` but that didn't really seem sensible?

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Oh, I missed that this is moved in from there, sorry.

But then why the changes in the `Alloc` struct, the changed doc comment for `bytes` and the new `size` field?

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Probably missed them when I cherry-picked some of the accessor utilities from the other PR.. Just now realized that they include references to not allocating all of the memory. Sorry for the confusion.

File: src/librustc/mir/interpret/allocation.rs
Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: That part of the doc shouldn't yet be there, that's part of the changeset for eliding allocation which was intended as a follow up instead :/

**Comment by HeroicKatora** [comment]
Removed `crate` visibility qualifier, only required rather simple changes to the stable hashing. Also removed the portion of comments on `bytes` and `undef_mask` that were not intended to be part of this PR yet.

**Comment by HeroicKatora** [comment]
The `size` field makes it much simpler to access the allocation extent as a `Size`, avoiding many additional conversions from the `usize` length of the byte vector. So it might be worth keeping?

File: src/librustc/ich/impls_ty.rs
Comment (RalfJung) [inline src/librustc/ich/impls_ty.rs]: Hm, the reason we use a pattern here is to be alerted when new fields are added, so that we don't forget to also hash them.

@oli-obk is there any way to keep the fields private while having reliable stable hashing? Can we derive this, or move the impl to `allocation.rs`?

Comment (RalfJung) [reply src/librustc/ich/impls_ty.rs]: `#[derive(HashStable)]` exists but I'll leave it to @oli-obk whether we can use that for `Allocation`.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    /// Denotes which part of this allocation is initialized.
```

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    /// The size of the allocation. Currently, must always equal `bytes.len()`.
```

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: `get_bytes_with_undef_and_ptr` also does not do bounds checks.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: Could you move the two methods relating to this down here, below `struct AllocationDefinedness`, into their own `impl` block?

File: src/librustc/ty/print/pretty.rs
Comment (RalfJung) [inline src/librustc/ty/print/pretty.rs]: Please add a comment explaining why using `raw` is okay here.

File: src/librustc/ty/print/pretty.rs
Comment (RalfJung) [inline src/librustc/ty/print/pretty.rs]: Please add a comment explaining why using `raw` is okay here.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Also please extend the doc comment to say something like "every use of this method should come with justification for why `get_bytes_with_undef_and_ptr` wasn't used".

File: src/librustc_codegen_llvm/consts.rs
Comment (RalfJung) [inline src/librustc_codegen_llvm/consts.rs]:
```suggestion
                        //
                        // The `raw` method is okay here because we checked relocations, and because we are doing
                        // this access to inspect the interpreter state (not as part of the interpreter execution).
                        alloc.raw_bytes_with_undef_and_ptr(all_bytes_range)
```

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Probably it should also say that this should only be used when inspecting the interpreter state "from the outside", not for interpreter execution. Maybe that can be reflected in the name?

File: src/librustc_codegen_llvm/consts.rs
Comment (RalfJung) [inline src/librustc_codegen_llvm/consts.rs]: This `alloc.relocations.len() == 0` is new, right? I was told before that this is already inside a section that checks that there are no relocations. So maybe make this an assertion instead?

File: src/librustc_mir/interpret/snapshot.rs
Comment (RalfJung) [inline src/librustc_mir/interpret/snapshot.rs]: Same concern with the `..` pattern here.

Maybe add a comment at the `Allocation` type like
```
// NOTE: When adding new fields, make sure to adjust the StableHash impl in ... and the Snapshot impl in ...
```

**Review by RalfJung — COMMENTED** [review]
This looks good overall, my notes are mostly editorial / ask for more comments.

Note that I'll be on conference next week and won't have time for reviews.

File: src/librustc/ich/impls_ty.rs
Comment (oli-obk) [reply src/librustc/ich/impls_ty.rs]: If we implement `HashStable` for `relocations` we should be able to use the derive. Or we just move this impl to `librustc/mir/interpret/allocation.rs`

File: src/librustc_codegen_llvm/consts.rs
Comment (oli-obk) [inline src/librustc_codegen_llvm/consts.rs]: use an `assert` here

File: src/librustc/mir/interpret/allocation.rs
Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Maybe `dump_bytes_with_undef_and_ptr` or `uninterpreted_bytes_with_undef_and_ptr`?

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: `dump` sounds like a debugging thinking dump to stderr.

`uninterpreted` doesn't really convey anything `raw` doesn't convey. `meta_level_inspect_bytes_with_undef_and_ptr` would work for me but I doubt it works for many others here.^^

Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: I'm all for verbose names when that definitely clarifies it. Better than having a method name and misinterpreting what it means because no one reads docs

Comment (Centril) [reply src/librustc/mir/interpret/allocation.rs]: > because no one reads docs

@oli-obk  I will have you that I definitely read docs (with a convenient firefox key `rustc`) when doing reviews where I haven't seen a method. =)

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: @RalfJung You're right, those aren't much better. I would find `meta` to be too generic, my first thought would be that it returns a property *of the bytes* rather than returning a raw part of the interpreter state. It works somewhat similar to a debugger attached to the interpreter, doesn't it? Maybe `inspect_internal_raw_bytes_with_undef_and_ptr` makes it clear that this is not to be used as part of interpreter execution.

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: I was thinking `inspect` as well (not sure `internal` adds anything), but one could reasonably say that the machine also "inspects" its bytes.

`inspect_from_outside` or so... maybe?^^

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Combining parts of the above: `inspect_bytes_with_undef_and_ptr_outside_interpreter`?

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Well, that's really long. But it does convey the crucial point.

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Maybe the `_bytes`-part could be dropped. 'inspect' and 'with_undef_and_ptr' already make it clear which part of the allocation is returned. That drops it to slightly below half the line length.

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: So that would be `inspect_with_undef_and_ptr_outside_interpreter`. @oli-obk is that excessive?

Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: No, not excessive, it makes it very clear that something nontrivial is going on and gives a good hint as to what that thing is.

File: src/librustc/mir/interpret/allocation.rs
Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: There already exists a method `relocations`, which returns relocations within a range specified by `ptr` and `size`, so I'd name the corresponding accessor `all_relocations` instead.

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Also, `interpret/memory.rs` tries to access it directly. I'd move code in a similar manner as compressed undef mask.

**Comment by HeroicKatora** [comment]
Moved the `HashStable` implementation into the module itself, renamed the raw accessor method, added comments for calls to it.

**Comment by RalfJung** [comment]
(I'm on a conference for the next week so won't have time for reviews, sorry.)

File: src/librustc_mir/interpret/snapshot.rs
Comment (oli-obk) [inline src/librustc_mir/interpret/snapshot.rs]: It's OK because we only use it for detecting loops and we extract the undef mask and relocations separately

Comment (HeroicKatora) [reply src/librustc_mir/interpret/snapshot.rs]: Good to know, included that in the reasoning.

File: src/librustc/mir/interpret/allocation.rs
Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: Sounds good

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: I'm not available until Wednesday next week but the rest should be ready.

Comment (oli-obk) [reply src/librustc/mir/interpret/allocation.rs]: take all the time you need

File: src/librustc_codegen_llvm/consts.rs
Comment (RalfJung) [inline src/librustc_codegen_llvm/consts.rs]:
```suggestion
            // This `inspect` is okay since we have checked that it is not within a relocation, it is
```

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    /// Looks at a slice which may describe undefined bytes or describe a relocation. This differs
```

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    /// Returns the undef mask.
```

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: `compress_undef_range` better reflects that this is about `undef_mask` IMO.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: What about `mark_compressed_undef_range`?

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: Good point, adding `undef` into the name.

File: src/librustc/mir/interpret/allocation.rs
Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: The name `fn relocations` is already taken: https://github.com/rust-lang/rust/blob/85ed538d6988c6c82aea8750b306cb793e874294/src/librustc/mir/interpret/allocation.rs#L424-L430.

I'd rename the existing method to `get_relocations` which would be aligned with `get_bytes`, `check_bytes`, `check_relocations` to be concerned with interpreter execution.

**Comment by HeroicKatora** [comment]
The `relocations` field is now also private, the direct access in `interpret/memory.rs` has been moved into the module similar to the previous change for `undef_mask`.

**Comment by HeroicKatora** [comment]
Open to feedback on the naming especially since I'm not perfectly content with it either.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Sounds good.
Makes me wonder who will even need the general read-only general accessor. Maybe that should also be `raw_relocations` or so, to make it less appealing to call?

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: Interesting idea. Long-term (does not have to happen in this PR), it might make sense to combine this with the run-length encoded undef mask into one struct? They are AFAIK only used together, and basically together opaquely represent "metadata" (on top of the raw bytes) that memcpy needs to copy. Then there would be two operations, tentatively named `get_metadata` and `apply_metadata` or so.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: Indentation looks off here.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: What about `get_relocations_range` as name?

**Review by RalfJung — APPROVED** [review]
LGTM, with the ntis fixed! However, I can't tell whether the arguments made in `codegen_llvm` are correct. Handing over to 

r? @oli-obk

**Comment by RalfJung** [comment]
r? @oli-obk 

(Doesn't look like the bots looks into reviews.)

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: `inital` might be a better field name? Also please add a doc comment, and maybe reorder the field.

My first reaction when I just saw this "why does it matter if this is the first definedness every created...?"

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: Please add doc comment that this expects the previous relocations for that range to be already cleared.

File: src/librustc/mir/interpret/allocation.rs
Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: The return value is destination specific since it already offsets correctly. I didn't think `get` would make this sufficiently clear. It would be possible to move the final offsetting with `dest.offset into `mark_relocation_range` but I didn't want to modify it to stay as close to the current code as possible.

Comment (RalfJung) [reply src/librustc/mir/interpret/allocation.rs]: Hm, makes sense.

File: src/librustc/mir/interpret/allocation.rs
Comment (RalfJung) [inline src/librustc/mir/interpret/allocation.rs]: Tiny nit (but I thought of it and now I can't not say it^^): might make more sense to make `initial` the first field.
When someone sees this type, I think it is easier to understand if they see this one first.

Then the doc comment for `ranges` can just say "at the end of each range, the boolean state flips", or something like that.

Comment (HeroicKatora) [reply src/librustc/mir/interpret/allocation.rs]: I see the point, now the doc of `ranges` references `initial` and ready well in natural sequence order.

**Comment by HeroicKatora** [comment]
@oli-obk I fixed the remaining nits from Ralf now, there are some unresolved review comments where you commented and I wasn't certain they were fully addressed. And the review of codegen is still oustanding as mentioned. Status should now be waiting on review.

**Comment by oli-obk** [comment]
@bors r+

Oh, right. I reviewed those comments when you made them. Reviewed again, still all good.

**Pull request merged**
